nixos/modules/services/matrix/mautrix-meta.nix

```diff
@@ -0,0 +1,210 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.mautrix-meta;
+  fullDataDir = "/var/lib/${cfg.dataDir}";
+  relativeDataDir = cfg.dataDir;
+  pkg = cfg.package;
+  registrationFile = "${fullDataDir}/meta-registration.yaml";
+  settingsFormat = pkgs.formats.yaml {};
+  settingsFile = "${fullDataDir}/config.yaml";
+  settingsFileUnsubstituted =
+    settingsFormat.generate "mautrix-meta-config.yaml" cfg.settings;
+in {
+  options = {
+    services.mautrix-meta = {
+      enable = mkEnableOption (mdDoc "Mautrix-Meta, a Matrix <-> Facebook and Matrix <-> Instagram hybrid puppeting/relaybot bridge");
+
+      package = mkPackageOption pkgs "mautrix-meta" { };
+
+      dataDir = mkOption {
+        type = types.path;
+        default =  "mautrix-meta";
+        description = mdDoc ''
+          Path to the directory with database, registration, and other data for the bridge service.
+          This path is relative to `/var/lib`, it cannot start with `../` (it cannot be outside of `/var/lib`).
+        '';
+      };
+
+      settings = mkOption rec {
+        apply = recursiveUpdate default;
+        inherit (settingsFormat) type;
+        default = {
+          homeserver = {
+            software = "standard";
+          };
+
+          appservice = rec {
+            database = {
+              type = "sqlite3-fk-wal";
+              uri = "file:${fullDataDir}/mautrix-meta.db?_txlock=immediate";
+            };
+
+            hostname = "0.0.0.0";
+            port = 29319;
+            address = "http://localhost:${toString cfg.settings.appservice.port}";
+          };
+
+          meta = {
+            mode = "instagram";
```

Comment:
Do you have an idea on how to realize that? To my understanding since all the settings options are not defined precisely, only removing this option the default config from the bridge itself will be used. Maybe one approach would be to expose something like `services.mautrix-meta.mode` that would be required and would be propagated to settings as the default value. 

Since this bridge allows either instagram or Facebook, I agree that it would make sense to make it multi instance. But again, I am not sure about the best approach to this. Should there be an array for example? Or something like `services.mautrix-meta.instagram`, `mautrix-meta.facebook` where both will have the same configurafion options?

Reply:
> Maybe one approach would be to expose something like services.mautrix-meta.mode that would be required and would be propagated to settings as the default value

I would probably just add an assertion. That way something needs to be configured in Nix. Another option may to be to do something like what Tor does and define NixOS options for a subset of `settings`. This allows validation and improves docs for common settings. You can then say that this value is required.

Reply:
> I agree that it would make sense to make it multi instance. But again, I am not sure about the best approach to this.

I think I would take a look at the nixos Redis module. It allows configuring any number of instances. This is probably a bit overkill for this since you would probably want at most 1 Instagram and 1 Facebook messenger instance but once you have 2 supporting an arbitrary number is pretty easy.

Maybe we can even preconfigure something like `services.mautrix-meta.instance.facebook` for messenger (change default bot name and stuff) and `services.mautrix-meta.instance.instagram` for Instagram. Then all the user needs to do for the common case is `services.mautrix-meta.instance.facebook.enable = true`. However if they have more complex needs they can define their own instances (instead of or in addition to the preconfigured ones).

Reply:
Thank you for pointing out that redis uses multiple instances, I have tried to replicate something similar. I have also added assertions for the meta.mode, and also decided to add assertions for homeserver domain and address, since the bridge won't start without these, and I don't think we can configure default domain.

Reply:
Multiple instances are now possible, and tested in a test. If there are any more issues, they will be discussed in other therads, resolving.